src/cpp/commandreply.cpp

```diff
@@ -30,236 +30,294 @@
-               <<this->_reply->integer<<std::endl;
-      break;
-    default:
-      std::cout<<index_tracker
-               <<" value type not supported."<<std::endl;
-  }
-  return;
+    //TODO these recursive functions can't use 'this' unless
+    //we have a constructor that takes redisReply*
+    std::cout<<index_tracker + " type: "
+             <<redis_reply_type()<<std::endl;
+    switch (_reply->type) {
+        case REDIS_REPLY_STRING:
+            std::cout<<index_tracker + " value: "
+                     <<std::string(str(),
+                                   str_len())
+                     <<std::endl;
+            break;
+        case REDIS_REPLY_ARRAY:
+            for (size_t i = 0; i < n_elements(); i++) {
+              std::string r_prefix = index_tracker + "[" +
+                                     std::to_string(i) + "]";
+              CommandReply tmp = (*this)[i];
+              tmp.print_reply_structure(r_prefix);
+            }
+            break;
+        case REDIS_REPLY_INTEGER:
+            std::cout<<index_tracker + " value: "
+                     <<_reply->integer<<std::endl;
+            break;
+        case REDIS_REPLY_DOUBLE:
+            std::cout<<index_tracker + " value: "
+                     <<_reply->dval<<std::endl;
+            break;
+        case REDIS_REPLY_ERROR:
+            std::cout<<index_tracker + " value: "
+                     <<std::string(str(), str_len())
+                     <<std::endl;
+            break;
+        case REDIS_REPLY_BOOL:
+            std::cout<<index_tracker + " value: "
+                     <<_reply->integer<<std::endl;
+            break;
+        default:
+            std::cout<<index_tracker
+                     <<" value type not supported."<<std::endl;
+    }
+}
+
+redisReply* CommandReply::deep_clone_reply(const redisReply* reply)
+{
+    if (reply == NULL)
+        return NULL;
+    redisReply* redis_reply = new redisReply;
+    *redis_reply = *reply;
+    redis_reply->str = NULL;
+    redis_reply->element = NULL;
+    switch (redis_reply->type) {
+        case REDIS_REPLY_ARRAY:
+        case REDIS_REPLY_MAP:
+        case REDIS_REPLY_SET:
+            // allocate memory for element and do deep copy
+            if (redis_reply->elements > 0) {
+                redis_reply->element = new redisReply*[redis_reply->elements];
+                if (reply->element != NULL) {
+                    for (size_t i = 0; i < reply->elements; i++)
+                        redis_reply->element[i] = deep_clone_reply(reply->element[i]);
+                }
+            }
+            break;
+        case REDIS_REPLY_ERROR:
+        case REDIS_REPLY_STATUS:
+        case REDIS_REPLY_STRING:
+        case REDIS_REPLY_DOUBLE:
+            // allocate memory for str and do deep copy
+            if (redis_reply->len > 0) {
+                redis_reply->str = new char[redis_reply->len];
+                std::strcpy(redis_reply->str, reply->str);
```

Comment:
I think we should do a memcpy here.  The string may contain null characters, and I think strcpy will terminate at the first null character.  We want to make sure we copy the whole string.